### PR TITLE
[FLINK-24475] Remove no longer used NestedMap* classes

### PR DESCRIPTION
**Pull request by Zakelly**
## What is the purpose of the change

Since NestedStateMap is no longer used in production code after FLINK-21935, this PR removes all NestedMap* classes.

## Brief change log

- Remove NestedStateMap, NestedMapsStateTable and NestedStateMapSnapshot
- Remove StateTableSnapshotCompatibilityTest since there is only one implement of StateTable.
- Change some tests 

## Verifying this change

This change is a code cleanup without any test coverage.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by Zakelly** [comment]
@flinkbot run azure

**Comment by Zakelly** [comment]
@Myasuka Would you please to take a look?

File: flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendTestBase.java [hunk 1]
```diff
@@ -4376,7 +4373,7 @@ protected void testConcurrentMapIfQueryable() throws Exception {
 
                 StateTable<?, ?, ?> stateTable =
                         ((AbstractHeapState<?, ?, ?>) kvState).getStateTable();
-                checkConcurrentStateTable(stateTable, numberOfKeyGroups);
+                assertNotNull("State not set", stateTable);
```
Comment (Myasuka) [inline flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendTestBase.java:4376]: I think `assertNotNull("State not set", ((AbstractHeapState<?, ?, ?>) kvState).getStateTable());` would be better.

**Review by Myasuka — COMMENTED** [review]
Thanks for the contribution, this PR looks good in general.

File: flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendTestBase.java (hunk 1 above)
Comment (Zakelly) [reply flink-runtime/src/test/java/org/apache/flink/runtime/state/StateBackendTestBase.java]: Sure, I've changed all three occurrences.

**Review by Myasuka — APPROVED** [review]

**Comment by Zakelly** [comment]
@Myasuka Thanks for the help.

**Comment by Myasuka** [comment]
The CI failed, @Zakelly do you have local green result?

**Comment by Zakelly** [comment]
@Myasuka Unfortunately no, the CI failed with an error code 137. I'll trigger another CI.

**Comment by Zakelly** [comment]
@flinkbot run azure

**Pull request closed without merging**